src/repo/mod.rs

```diff
@@ -235,8 +238,34 @@ impl TryFrom<RequestKind> for RepoEvent {
     }
 }
 
+#[derive(Debug)]
+struct Lock {
+    file: File,
+}
+
+impl Lock {
+    fn new(path: &Path, create: bool) -> std::io::Result<Lock> {
+        use std::fs::OpenOptions;
+
+        let file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(create)
+            .open(path)
+            .unwrap();
+
+        FileExt::try_lock_exclusive(&file)?;
+
+        Ok(Lock { file })
+    }
+}
+
 impl<TRepoTypes: RepoTypes> Repo<TRepoTypes> {
     pub fn new(options: RepoOptions) -> (Self, Receiver<RepoEvent>) {
+        let lockfile_path = options.path.join("repo_lock");
+        let create = !lockfile_path.is_file();
```

Comment:
I think this will be a TOCTOU issue, which could be worked around by looping on creating and opening for writing but ... I am not entirely sure. Perhaps this is ok for initial attempt.

Reply:
We could actually remove this check entirely as `OpenOptions::create` will create if it doesn't exist and simply open if it does. I *think* that would remove the TOCTOU issue?

Reply:
Yeah I think it will take care of this.

I was initially thinking that a loop of "create new" and "open existing for write" was needed but can't see how it would be better than just a single "create if not exist" file open. The important thing is not to ever create a new file to replace the old, which I think is not possible with plain open, it'd need remove as well so that's not something done by mistake :)